spanner/ot_metrics.go

```diff
@@ -0,0 +1,258 @@
+		metric.WithDescription("Number of RPC responses received without the server-timing header, most likely means that the RPC never reached Google's network"),
+		metric.WithUnit("1"),
+	)
+	if err != nil {
+		logf(logger, "Error during registering instrument for metric spanner/gfe_header_missing_count, error: %v", err)
+	}
+	config.gfeHeaderMissingCount = gfeHeaderMissingCountInstrument
+}
+
+func registerSessionPoolOTMetrics(pool *sessionPool) error {
+	otConfig := pool.otConfig
+	if !IsOpenTelemetryMetricsEnabled() || otConfig == nil {
+		return nil
+	}
+
+	attributes := otConfig.attributeMap
+	attributesInUseSessions := append(attributes, attributeNumInUseSessions)
+	attributesAvailableSessions := append(attributes, attributeNumSessions)
+
+	reg, err := otConfig.meterProvider.Meter(OtInstrumentationScope, metric.WithInstrumentationVersion(internal.Version)).RegisterCallback(
+		func(ctx context.Context, o metric.Observer) error {
+			pool.mu.Lock()
+			defer pool.mu.Unlock()
+			log.Print("RegisterCallback called for meter")
+
+			o.ObserveInt64(otConfig.openSessionCount, int64(pool.numOpened), metric.WithAttributes(attributes...))
+			o.ObserveInt64(otConfig.maxAllowedSessionsCount, int64(pool.SessionPoolConfig.MaxOpened), metric.WithAttributes(attributes...))
+			o.ObserveInt64(otConfig.sessionsCount, int64(pool.numInUse), metric.WithAttributes(attributesInUseSessions...))
+			o.ObserveInt64(otConfig.sessionsCount, int64(pool.numSessions), metric.WithAttributes(attributesAvailableSessions...))
+			o.ObserveInt64(otConfig.maxInUseSessionsCount, int64(pool.maxNumInUse), metric.WithAttributes(attributes...))
+
+			return nil
+		},
+		otConfig.openSessionCount,
+		otConfig.maxAllowedSessionsCount,
+		otConfig.sessionsCount,
+		otConfig.maxInUseSessionsCount,
+	)
+	pool.otConfig.otMetricRegistration = reg
+	return err
+}
+
+// EnableOpenTelemetryMetrics enables OpenTelemetery metrics
+func EnableOpenTelemetryMetrics() {
+	setOpenTelemetryMetricsFlag(true)
+}
+
+// IsOpenTelemetryMetricsEnabled tells whether OpenTelemtery metrics is enabled or not.
+func IsOpenTelemetryMetricsEnabled() bool {
+	otMu.RLock()
+	defer otMu.RUnlock()
+	return openTelemetryMetricsEnabled
+}
+
+func setOpenTelemetryMetricsFlag(enable bool) {
+	otMu.Lock()
+	openTelemetryMetricsEnabled = enable
+	otMu.Unlock()
+}
+
+func recordGFELatencyMetricsOT(ctx context.Context, md metadata.MD, keyMethod string, otConfig *openTelemetryConfig) error {
+	if !IsOpenTelemetryMetricsEnabled() || md == nil && otConfig == nil {
+		return nil
+	}
+	attr := otConfig.attributeMap
+	if len(md.Get("server-timing")) == 0 && otConfig.gfeHeaderMissingCount != nil {
+		otConfig.gfeHeaderMissingCount.Add(ctx, 1, metric.WithAttributes(attr...))
+		return nil
+	}
+	serverTiming := md.Get("server-timing")[0]
+	gfeLatency, err := strconv.Atoi(strings.TrimPrefix(serverTiming, "gfet4t7; dur="))
+	if !strings.HasPrefix(serverTiming, "gfet4t7; dur=") || err != nil {
+		return err
```

Comment:
It's a bit of a theoretical case, but shouldn't we split this into two separate things:
1. `err != nil`: return the error
2. `!strings.HasPrefix(serverTiming, "gfet4t7; dur=")`: This should be considered the same as a missing GFE header (or malformed GFE header). Now we just return `nil` when that happens, is that intentional?

Reply:
In case of malformed GFE header we return an error, not `nil`. 
We are not counting malformed GFE as missing GFE because in this case we get a GFE header but it is not in required format.

Reply:
ack

note that this could still return `err=nil`, if the actual text is only a number without any other prefixes or suffixes. The reason is that `strings.TrimPrefix` returns the original string if the string did not have the prefix.

This for example prints `4`:

```go
fmt.Println(strconv.Atoi(strings.TrimPrefix("4", "World")))
```